Support parent relative paths in Path.relative_to

pnezis opened the pull request:
Adds a `:force` option in `Path.relative_to` and `Path.relative_to_cwd` which allows up traversals.


## Initial PR body

`Path.relative_to` does not return proper relative paths in some cases. Some examples follow (also check the added tests):

```elixir
# If two relative paths are given where from is a subfolder of path
# it returns the path unchanged 
iex> Path.relative_to("foo", "foo/bar")
"foo" # wrong, implies ./foo/bar/foo - should be ../foo

# Paths are not properly expanded, all following should return "usr"
iex> Path.relative_to("/usr/../usr", "/")
"usr/../usr"

iex> Path.relative_to("/usr", "/etc/..")
"/usr"

iex> Path.relative_to("/./etc/../usr", "/")
"./etc/../usr"

# Relative paths to parent folders are never returned
iex> Path.relative_to("/usr/foo/tmp", "/usr/foo/bar/buzz")
"/usr/foo/tmp" # Should be ../../tmp
```

By expanding first the two paths we can easily calculate the number of "cd ups" and construct a relative path. The only case for which an absolute path should be returned is when it is not possible to `mv` from `from` to `path`. I think this can only happen in Windows when the two paths are in different drives, but I don't have access to a Windows machine to properly test it.

I found a similar implementation in `Mix.Utils` which I replaced with a call to `Path.relative_to` in a separate commit.~~

josevalim commented:
I am not sure we should do this change because it will implicitly assume that both paths belong to the same cwd, which may not necessarily be true, for example, I may want this:

    Path.relative_to "baz",  "foo/bar"

to simply return `baz` because there is no common element. Or I may want `(/usr, /etc)` to simply return `/utc`.

I understand the feature you want though and it is a valid feature but I think changing the implementation here is most likely a breaking change, given by the fact it also changes the function documentation. So I propose we define a new function that does the operation you desire (and use it in Mix too). :)

pnezis commented:
Agreed, any suggestion for the name (`relative_path` maybe like pyhton's `relpath`)? Should `relative_to_cwd` use the new one since it implies we are on the cwd?

josevalim commented:
`relative_to_cwd` implies we are on `cwd` but it doesn't imply I want non-relative paths. For example, we use it in error messages. If a path is not on cwd, do we want to show:

    ../../../baz/bat.ex

or:

    /Users/jose/OSS/baz/bat.ex

I think the second is much clearer. :)

I am not sure about the name. Calling it relative will be confusing but perhaps we can have an additional optional argument on `relative_to`?  But also I am not sure how to call it. We could call it `expand` but it does more, it also allows "up-traversals".

pnezis commented:
Putting aside the "up-traversals", I agree that the proposed change is likely breaking, but I find the current implementation somewhat inconsistent.

> I am not sure we should do this change because it will implicitly assume that both paths belong to the same cwd, which may not necessarily be true, for example, I may want this:
> 
> ```
> Path.relative_to "baz",  "foo/bar"
> ```
> 
> to simply return `baz` because there is no common element. Or I may want `(/usr, /etc)` to simply return `/utc`.

When the arguments are relative paths it should be implicitly assumed that both paths belong to the same `cwd`, otherwise why would I expect

```
Path.relative_to "foo/bar/baz" "foo/bar"
```

to return `baz`?

For me the following invariant should hold for a function that returns the relative path:

```
assert join(from, relative_path(path, from)) == path
``` 

The inconsistencies of the current implementation originate from the fact that relative path arguments are not expanded (the `cwd` upon which they are expanded does not matter as long as it is the same for both paths) .

For example:

```elixir
File.cd!("/usr/local") # set cwd

Path.relative_to("/usr/foo", "/usr") # foo -OK
Path.relative_to("/usr/foo", "/etc") # /usr/foo -OK 
Path.relative_to("foo/bar", "foo") # bar -OK
Path.relative_to("bar", "foo"), # bar - NOK, I would expect /usr/local/bar to be returned in this case
Path.relative_to("/usr/local/foo", "."), # /usr/local/foo - OK but if "." was expanded it would return foo 
Path.relative_to("foo/../foo/bar/../bar/bat.ex", "foo/bar"), # foo/../foo/bar/../bar/bat.ex - OK if it was expanded it would return bat.ex

# similarly for relative_to_cwd. Notice that here the returned path is always either a valid relative path or a valid absolute
# path since the second argument is guaranteed to be an absolute path
Path.relative_to_cwd("foo") # foo - OK
Path.relative_to_cwd("/usr/local/foo") "foo - OK"
Path.relative_to_cwd("../local/foo") # "../local/foo" - OK but if it was expanded it would return "foo"
Path.relative_to_cwd("/etc/../usr/local/foo") # "/etc/../usr/local/foo" - again it would return "foo" if expanded 
```
- Should we update the docs to indicate that in case of relative paths the returned path is not guaranteed to be valid relative path?
- Should we change `Path.relative_to_cwd` to expand the `path` before calling `relative_to`? 

## Suggestion

Add a new `Path.relative_path/2` (or any better name you can think of :)) that implements the suggested change.  

Similar functions in other languages:

- Python: [`os.path.relpath`](https://docs.python.org/3/library/os.path.html#os.path.relpath)
- C++ : [`relative`](https://en.cppreference.com/w/cpp/filesystem/relative)
- Node: [`path.relative`](https://nodejs.org/api/path.html#path_path_relative_from_to)

@josevalim Thanks for the feedback, should I move the discussion to the mailing list?

josevalim commented:
> When the arguments are relative paths it should be implicitly assumed that both paths belong to the same cwd, otherwise why would I expect to return `baz`?

You are right. That was a bad argument. But I think the point still stands that we don't want to necessarily perform up-traversals (as an example, the several places we use this in the Elixir compiler).

> Path.relative_to("foo/bar", "foo") # bar -OK
> Path.relative_to("bar", "foo"), # bar - NOK, I would expect /usr/local/bar to be returned in this case

Right, it seems we may have incorrect behaviour when both arguments are relative paths. There are two possible interpretations:

1. Assume the second argument is the `cwd` for the first argument, so example 2 above is correct and 1 is wrong
2. Assume they both belong to a shared `cwd`, which means example 1 is correct and 2 is wrong

All examples use absolute paths, which may indicate that this function never really expected to receive relative paths. I am still worried this could somehow be a breaking change :( I am also not convinced we should _expand_ (it can be done by callers before hand).

Here is my suggestion:

1. We must warn if the second argument is a relative path
2. If the first argument is relative, we simply return it (i.e. we assume the path is relative to the second argument which is the current behaviour)
3. If they are both absolute it, behave as is
4. Introduce a `:traverse_up` option that triggers your proposed behaviour when they are both absolute

The user is responsible to expand/normalize it upfront if the want to. Implementation (pseudo-code) would be something like:

```elixir
second_arg_type = Path.type(second_arg)
traverse_up = Keyword.get(opts, :traverse_up, false)

if second_arg_type == :relative do
  IO.warn "..."
end

cond do
  not traverse_up? ->
    current_implementation(...)

  Path.type(first_arg) != :absolute ->
    first_arg

  true ->
    your_new_implementation(...)
end
```

WDYT? The goal is to minimize the amount of deprecations while keeping the non-invalid behaviour we have today.

I think this is a bug, so we should solve it, no need for proposal. thank you for guiding me here.

pnezis commented:
Sounds like a plan, I would only emit the warning tough if the `traverse_up` option is not set. I will update the PR with the proposed changes if you agree.

pnezis commented:
> Sounds like a plan, I would only emit the warning tough if the `traverse_up` option is not set. I will update the PR with the proposed changes if you agree.

Ignore it, I missed the not-expansion by default part

pnezis commented:
@josevalim the only problem I see with not expanding by default is that you will still not get "correct" results, e.g.

```
Path.relative_to("/usr/local/../", "/usr", traverse_up: false) # would return local/... instead of .
Path.relative_to("/usr/../etc", "/etc",  traverse_up: true) # would return ../usr/../etc
Path.relative_to("/usr", "/etc/..", traverse_up: true) # would return ../../usr which is incorrect
Path.relative_to("/usr", "/usr/..", traverse_up: true) # would return ".."
```

`Path.type` marks all the above paths as `:absolute`. If `traverse_up` is set then the paths should be expanded otherwise such inconsistencies will incur.

pnezis commented:
So to recap the plan is the following:

**Current behaviour**

- If the paths are absolute always expand, this will fix some issues present today, e.g. `Path.relative_to("/usr/local/..", "/usr"`) would return `"."` instead of `"local/.."`, without introducing any breaking change
- If the second argument is relative, issue a warning
- Update the docs accordingly

**New `traverse_up` option**

- Add an extra `opts` argument to `Path.relative_to` and `Path.relative_to_cwd` with a `:traverse_up` option, that defaults to `false`
- If set the paths are always expanded (both relative and absolute) and the new implementation is called.
- Update the docs accordingly

 If you are fine with the proposed changes I will update the PR accordingly

pnezis commented:
> Yes, I think those are correct for the given paths. If you want to address them, then go ahead and expand it before. :) 

It is not correct in all cases. Check the sample code below, it's the new implementation with expansion optional:

```elixir
defmodule RelativePath do
  def relative_to(path, start, opts \\ []) do
    path = 
      path
      |> maybe_expand(Keyword.get(opts, :expand, true))
      |> Path.split()
    
    start = 
      start
      |> maybe_expand(Keyword.get(opts, :expand, true))
      |> Path.split()
    
    do_relative_to(path, start, path)
  end

  defp maybe_expand(path, true), do: Path.expand(path)
  defp maybe_expand(path, false), do: path

  defp do_relative_to([h | t1], [h | t2], path), do: do_relative_to(t1, t2, path)
  defp do_relative_to([], [], _path), do: "."

  defp do_relative_to(path, from, _path) do
    base = List.duplicate("..", length(from))
    Path.join(base ++ path)
  end

  def compare(path, start) do
    %{
      path: path,
      start: start,
      current_relative_to: check(path, start, Path.relative_to(path, start)),
      new_relative_to_no_expand: check(path, start, relative_to(path, start, expand: false)),
      new_relative_to_expand: check(path, start, relative_to(path, start, expand: true))
    }
  end

  defp check(path, from, relative) do
    reconstructed_path =
      case Path.type(relative) do
        :absolute -> Path.expand(path)

        :relative -> Path.join(from, relative) |> Path.expand()
      end

    valid? = Path.expand(path) == reconstructed_path

    {valid?, relative}
  end
end
```

Now try it with the following inputs:

```elixir
iex> RelativePath.compare("/usr", "/usr/..")
%{
  current_relative_to: {true, "/usr"},
  new_relative_to_expand: {true, "usr"},
  new_relative_to_no_expand: {false, ".."},
  path: "/usr",
  start: "/usr/.."
}
```

as you see with no expansion we get an invalid relative path.

Regarding the symlinks I agree but the function's documentation already mentions it:

> This function does not query the file system, so it assumes no symlinks between the paths.

josevalim commented:
Got it, thanks! So if we `traverse_up`, we need to have arguments expanded. That's clear now.

---

I am really struggling with this API. Here is what we know:

1. Documentation only uses absolute paths in examples

2. We want to add the capability of "traversing up" but the compiler usage does not want up traversals

3. Our main usage is with `Path.relative_to_cwd` which by definition has the second argument always expanded

So let's talk about the second argument first. We have two options here:

1. Warn when the second argument is a relative path and otherwise _assume_ the path has been expanded

2. Automatically expand the path. Pro: always correct. Cons: may be backwards incompatible. Cons: we may perform double work in cases like `Path.relative_to_cwd`

Which option do we want to take here?

pnezis commented:
> Got it, thanks! So if we `traverse_up`, we need to have arguments expanded. That's clear now.

My fault for not making it clear from the beginning

> 1. Warn when the second argument is a relative path and otherwise _assume_ the path has been expanded

I wouldn't go with this option, since when `traverse_up: true` it could give incorrect results and could be a potential security vulnerability (file system traversal). 

> 2. Automatically expand the path. Pro: always correct. Cons: may be backwards incompatible. Cons: we may perform double work in cases like `Path.relative_to_cwd`

If the path is expanded by default only if `traverse_up: true` then there would be no backwards incompatibility. Only redundant expansion if called through `Path.relative_to_cwd`, but It's something we can live with. But I don't like it either because the function is not consistent (conditional implicit expansion).

3. Do not expand by default but add a check that the second argument is expanded (only if `traverse_up: true`). If it is not expanded raise to be safe.

4. **Conditionally expand**. Introduce a second option `:expand` which triggers expansion for **both arguments** if set. If it is not set to `true` you still have to check that the second argument is expanded in case  `traverse_up: true`.  An extra advantage of this option is that it will also work correctly with relative paths (both with and without `traverse_up`). The API would look like this:

```elixir
Path.relative_to("foo," "foo/bar")     # foo - like now with a warning for foo/bar being relative
Path.relative_to("/usr/local", "/usr") # /local - like now nothing changes
Path.relative_to("/usr", "/usr/local") # /usr - like now no traverse_up is set
Path.relative_to("/usr", "/usr/local", traverse_up: true) # .. - all good second argument is expanded
Path.relative_to("/usr", "/usr/..", traverse_up: true) # raise - not expanded second argument
Path.relative_to("/usr", "/usr/..", traverse_up: true, expand: true) # usr - expansion and correct calculation
Path.relative_to("foo", "foo/bar", traverse_up: true, expand: true) == ".." - works with relative paths as well
Path.relative_to("foo", "bar", expand: true) # /full/path/to/foo
Path.relative_to("foo", "foo/bar", expand: true) # /full/path/to/foo - no traverse up is set
``` 

5. Do not add the `traverse_up` option in current implementations. Instead add a new function that expands both arguments by default.

What do you think? I would go with one of 3,4.

PS. The `expanded?` check could be exposed as a public function.

josevalim commented:
We could add another option but then we end-up exposing more to the user, having more complexity on the user facing API (which we have to balance with backwards compatibility).

Here is one additional idea:

1. Let's separate the path expansion to an absolute path and the expansion of dots as two separate operations. The second one is cheap and does not require OS traversals

2. If any of the paths are relative, we return the first argument as is (we can potentially warn if the second argument is relative - and raise in the future)

3. Now we always expand the dots of both sides (cheap and fast)

4. The `traverse_up` option controls only the traversing up

josevalim commented:
Gah, I also just learned that we have `Path.safe_relative_to` which treats the second argument as `cwd` and expand paths, so we may want to synchronize the behaviors. I will submit a new proposal soon.

josevalim commented:
Hi @pnezis, here is a proposal: https://github.com/elixir-lang/elixir/pull/12632

As mentioned above, that PR decouples the absolute expansion from the dot-expansion and brings the behaviour closer to `safe_relative` and `safe_relative_to`. This means:

1. the second argument must be an absolute path (otherwise we warn and keep the previous behaviour for backwards compatibility)

2. if absolute paths are given, they are always dot-expanded

3. if relative paths are given, they are always dot-expanded

### The traverse-up option

I think this allows us to add `:traverse_up` without a lot of mental gymnastics. The docs could say: "By default, no up traversals are added beyond the ones in the original path. With this option set to true, we add any necessary up traversals to force a relative path to be returned". Another option is to call the option `:force` so we say: "Force a relative path to be returned by traversing up. Except if the paths are in different volumes on Windows". At this moment, I am inclined to call it `:force`.

pnezis commented:
Hi @josevalim. +1 for the proposal, it is cleaner compared to the extra argument and addresses all of the issues we discussed.

I also prefer `:force` to `:traverse_up`. Once #12632 is merged I will update this accordingly. Thanks a lot!

josevalim commented:
@pnezis merged, thanks for all the help. So let's go with the `:force` option and remember that we cannot traverse up if we are on Windows on different drives. :)

pnezis commented:
Hi @josevalim I don't have time to work on it today but I will have it ready tomorrow

>  remember that we cannot traverse up if we are on Windows on different drives. :)

I won't forget :) , I addressed this also in the initial PR

pnezis commented:
Hi @josevalim , PR updated. I don't have access to a Windows machine to test let's see if the test suite passes.

josevalim commented on lib/elixir/lib/path.ex:
```diff
@@ -327,6 +333,12 @@ defmodule Path do
       Path.relative_to("/usr/local/foo", "/usr/local/foo")  #=> "."
       Path.relative_to("/usr/local/../foo", "/usr/foo")     #=> "."
       Path.relative_to("/usr/local/../foo/bar", "/usr/foo") #=> "bar"
+  
```
```suggestion

```

josevalim commented on lib/elixir/test/elixir/path_test.exs:
```diff
@@ -191,11 +221,19 @@ defmodule PathTest do
+    splitted_cwd = Path.split(File.cwd!())
+    relative_to_root = List.duplicate("..", length(splitted_cwd) - 1)
+
+    assert Path.relative_to_cwd("/") == "/"
+    assert Path.relative_to_cwd("/", force: true) == Path.join(relative_to_root)
```
These assertions are Unix specific :D

pnezis commented on lib/elixir/lib/path.ex:
```diff
@@ -361,14 +374,22 @@ defmodule Path do
         split_path = expand_split(split_path)
         split_cwd = expand_split(split_cwd)
-        relative_to(split_path, split_cwd, split_path)
+        relative_to(split_path, split_cwd, split_path, force)
```
Sure any naming suggestion? `defp relative_to/3` will need to be renamed since it will match the public `relative_to` function

josevalim replied:
Any name is fine :) it is all private anyway. whatever makes you happy!

josevalim replied:
That will still work if the new argument is an optional argument (which is the case here!).

josevalim commented on lib/elixir/test/elixir/path_test.exs:
```diff
@@ -87,25 +87,36 @@ defmodule PathTest do
-      assert Path.relative_to("d:/usr/local/foo/..", "d:/usr/local/..") == "local"
-      assert Path.relative_to("d:/usr/local/foo/..", "d:/usr/local/.") == "."
+      # on different volumes with force: true it should return the original path
+      assert Path.relative_to("d:/usr/local", "c:/usr/local", force: true) == "d:/usr/local"
```
```suggestion
      assert Path.relative_to("d:/usr/local", "c:/usr/local", force: true) == "d:/usr/local"
      assert Path.relative_to("d:/usr/local", "c:/another/local", force: true) == "d:/usr/local"
```

pnezis commented on lib/elixir/lib/path.ex:
```diff
@@ -327,6 +333,11 @@ defmodule Path do
       Path.relative_to("/usr/local/foo", "/usr/local/foo")  #=> "."
       Path.relative_to("/usr/local/../foo", "/usr/foo")     #=> "."
       Path.relative_to("/usr/local/../foo/bar", "/usr/foo") #=> "bar"
+  If `:force` is set to `true` paths are traversed up:
```
Don't we need an empty line above this one?

josevalim commented on lib/elixir/lib/path.ex:
```diff
@@ -327,6 +333,11 @@ defmodule Path do
       Path.relative_to("/usr/local/foo", "/usr/local/foo")  #=> "."
       Path.relative_to("/usr/local/../foo", "/usr/foo")     #=> "."
       Path.relative_to("/usr/local/../foo/bar", "/usr/foo") #=> "bar"
+  If `:force` is set to `true` paths are traversed up:
```
Yeah, I removed the trailing space and accidentally the whole line with it.
